include/sempr/entity/spatial/AbstractPointCloud.hpp

```diff
@@ -0,0 +1,166 @@
+        iterator operator++() { iterator i = *this; index_++; return i; }
+        iterator operator++(int junk) { index_++; return *this; }
+        T& operator*() { return (*ptr_)[index_]; }
+        T* operator->() { return &operator*(); }
+        bool operator==(const iterator& rhs) { return ptr_ == rhs.ptr_ && index_ == rhs.index_; }
+        bool operator!=(const iterator& rhs) { return ptr_ != rhs.ptr_ || index_ != rhs.index_;; }
+    private:
+        AbstractChannel* const ptr_;
+        std::size_t index_;
+    };
+
+    class const_iterator
+    {
+    public:
+        const_iterator(const AbstractChannel* ptr, std::size_t index = 0) : ptr_(ptr), index_(index) { }
+        const_iterator operator++() { const_iterator i = *this; index_++; return i; }
+        const_iterator operator++(int junk) { index_++; return *this; }
+        const T& operator*() { return (*ptr_)[index_]; }
+        const T* operator->() { return &operator*(); }
+        bool operator==(const const_iterator& rhs) { return ptr_ == rhs.ptr_ && index_ == rhs.index_; }
+        bool operator!=(const const_iterator& rhs) { return ptr_ != rhs.ptr_ || index_ != rhs.index_;; }
+    private:
+        const AbstractChannel* ptr_;
+        std::size_t index_;
+    };
+
+    const_iterator cbegin() { return const_iterator(this, 0); };
+    const_iterator begin() const { return const_iterator(this, 0); };
+    iterator begin() { return iterator(this, 0); };
+
+    const_iterator cend() { return const_iterator(this, size()); }; 
+    const_iterator end() const { return const_iterator(this, size()); };
+    iterator end() { return iterator(this, size()); };
+};
+
+
+/**
+ * The AbstractPointCloud is an interface that defines how to access the elements of a point cloud (channels and points)
+ * In future a PointCloud implementation have to fullfill this interface.
+ * 
+ * Note: For now there is only readable access to the points of the PointCloud. If you like to modify the point you have to grab them and put they into a new PointCloud object.
+ */
+template<typename T>
+class AbstractPointCloud
+{
+public:
+    using Ptr = std::shared_ptr<AbstractPointCloud>;
+
+    virtual bool hasChannel(int type) const = 0;
+
+    /**
+     * It isnt possible to declare a template to a virtual method.
+     * And C++ have no return value overloading so the the only this manual way to do.
+     * It shall look like:
+     * 
+     * template<typename D>
+     * typename std::enable_if<std::is_fundamental<D>::value>
+     * virtual AbstractChannel<D>& getChannel(int type) = 0;
+     */
+    virtual void getChannel(int type, AbstractChannel<int8_t>& channel) const = 0;
+    virtual void getChannel(int type, AbstractChannel<int16_t>& channel) const = 0;
+    virtual void getChannel(int type, AbstractChannel<int32_t>& channel) const = 0;
+    virtual void getChannel(int type, AbstractChannel<int64_t>& channel) const = 0;
```

Comment:
Yes indeed it shall look like this. But as long as C++ doesn't count the return type as part of the signature it will only cause 'cannot be overloaded' errors.

I also try to use template instead but C++ does not like virtual template functions ;-)